boa_engine/src/string.rs

```diff
@@ -314,18 +617,81 @@ impl Inner {
 /// pointer is kept, so its size is the size of a pointer.
 #[derive(Finalize)]
 pub struct JsString {
-    inner: NonNull<Inner>,
+    inner: Flag,
     _marker: PhantomData<Rc<str>>,
 }
 
+/// It maybe an index of [`CONSTANTS_ARRAY`], or a raw pointer of [`Inner`].
+/// Use the first bit as the flag.
+/// Detail: <https://en.wikipedia.org/wiki/Tagged_pointer>
+#[repr(transparent)]
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+struct Flag(NonZeroUsize);
+
+impl Flag {
+    #[inline]
+    fn new_heap(inner: *mut Inner) -> Self {
+        // Safety: We already know it's not null, so this is safe.
+        Self(unsafe { NonZeroUsize::new_unchecked(inner as usize) })
+    }
+
+    /// Set the first bit to 1, indicating that it is static. Store the index
+    /// in 1..63 bits.
+    #[inline]
+    const fn new_static(idx: usize) -> Self {
+        // Safety: We already know it's not null, so this is safe.
+        Self(unsafe { NonZeroUsize::new_unchecked((idx << 1) | 1) })
+    }
+
+    /// Check if the first bit is 1.
+    #[inline]
+    const fn is_static(self) -> bool {
+        self.0.get() & 1 == 1
+    }
+
+    /// # Safety
+    ///
+    /// It maybe a static string.
+    #[inline]
+    const unsafe fn get_heap_unchecked(self) -> *mut Inner {
+        self.0.get() as *mut _
+    }
+
+    /// # Safety
+    ///
+    /// It maybe a string allocated on the heap.
+    #[inline]
+    const unsafe fn get_static_unchecked(self) -> &'static str {
+        // shift right to get the index.
+        CONSTANTS_ARRAY[self.0.get() >> 1]
```

Comment:
```suggestion
        CONSTANTS_ARRAY.get_unchecked(self.0.get() >> 1)
```
We should use `get_unchecked(index)`, since the index operator always does a boundary check.